[jayllo-c] iP

jayllo-c opened the pull request:
# Destiny
> “Your mind is for having ideas, not holding them.” – David Allen ([source](https://dansilvestre.com/productivity-quotes))
Destiny frees your mind of having to remember things you need to do. It's,
- text-based
- easy to learn
- ~~FAST~~ SUPER FAST to use
All you need to do is,
1. download it from [here](https://github.com/jayllo-c/ip/releases/tag/A-UserGuide).
2. double-click it.
3. add your tasks.
4. let it manage your tasks for you 😉 
And it is **FREE**
Features:

- [x] Managing tasks
- [x] Managing deadlines
- [ ] Reminders (coming soon)
##
If you are a Java programmer, you can use it to practice Java too. Here's the `main` method:
```
public class Main {
    public static void main(String[] args) {
        Application.launch(MainApp.class, args);
    }
}
```

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

Commit:
docs/README.md: Tweak document template

Let's tweak the docs/README.md (which is used as the user guide)
to fit Duke better. Specifically,

1. mention product name in the title
2. mention adding a product screenshot and a product intro
3. tweak the flow to describe feature-by-feature

jayllo-c committed:
Level-0 commit

jayllo-c committed:
no message

jayllo-c committed:
no message

jayllo-c committed:
no message

jayllo-c committed:
no message

jayllo-c committed:
no message

jayllo-c committed:
no message

jayllo-c committed:
Side note, I didn't realise i had to commit A-TextUiTesting and A-Enums, so I didnt do the textUiTesting properly until this commit, and enums were used since the level-5 commit. Will take proper note of the correct commit sequence moving forward

jayllo-c committed:
Duke.java: Add loadData and saveData functions

jayllo-c committed:
Duke Class: Update position of saveData() function in main()

The duke.txt file is supposed to be updated for every change made to
the list. Hence, the saveData() function has been moved to reflect this.
Prior to this, the .txt file would only be updated at the end of the
interaction with the Duke chat box when they use the command "bye".

jayllo-c committed:
Deadline and Event class: use java.time.LocalDateTime

Deadline and event classes now save their times as
java.time.LocalDateTime objects.

This is to enable destiny to understand dates and times. As such,
destiny is now able to do the following:
* display the date and time in a standardized manner
* for events, ensure start time is a date and time before end time

jayllo-c committed:
Duke class: use OOP

Now uses instance of Duke class instead of running duke from a static
context. This is to make the code more object oriented.

jayllo-c committed:
Duke class: use OOP

Now uses instance of Duke class instead of running duke from a static
context. This is to make the code more object oriented.

jayllo-c committed:
Revert "Duke class: use OOP"

This reverts commit bcf2969dbb286b49bd7e48a65e5b0e6655a08a88.

jayllo-c committed:
Duke class: Extract functions into classes

Extracted the following from the duke class:
* loading and saving of tasks in the data file (Storage.java)
* making sense of the user command (Parser.java)
* the taskList, and operations that manipulate it (TaskList.java)

This was done to make the code more object oriented

jayllo-c committed:
Update automated tests

Changed the automated tests to adjust for the recent changes to
the code and how it responds to correct and incorrect inputs.

jayllo-c committed:
Move all main java files to duke package

Transferred all main java files for duke into the duke package.

This is to understand how to create and utilize java packages.

jayllo-c committed:
Merge branch 'branch-A-Packages'

jayllo-c committed:
Merge remote-tracking branch 'origin/add-gradle-support'

jayllo-c committed:
Switch to Gradle

Set up Gradle to build and run Duke using Gradle.

This was done to allow the automation of project builds subsequently for
this project.

jayllo-c committed:
Merge branch 'branch-A-Gradle'

jayllo-c committed:
Add JUnit Tests

Add JUnit tests to test the behavior of the code.
The current classes tested are as follows:
* Deadline
* Event

jayllo-c committed:
Merge branch 'branch-A-JUnit'

jayllo-c committed:
Follow the Coding Standard

Updated all main java files in the duke package to follow the coding
standard.

jayllo-c committed:
Add JavaDoc comments to code

Updated the code to include JavaDoc comments to follow the coding
standard.

jayllo-c committed:
Duke Class: Add find function

Added the find(String) function to the Duke class.

This was done to give users a way to find task(s) by searching for
a keyword.

jayllo-c committed:
Merge branch 'branch-Level-9'

jayllo-c committed:
Merge branch 'branch-A-CodingStandard'

# Conflicts:
#	src/main/java/duke/Duke.java

Conflicts were resolved.

jayllo-c committed:
Merge branch 'branch-A-JavaDoc'

# Conflicts:
#	src/main/java/duke/Duke.java
#	src/main/java/duke/Storage.java
#	src/main/java/duke/TaskList.java

jayllo-c committed:
Improve Coding Standard

Updated entire code to ensure checkstyle does not have a single error.

This was done to ensure the coding standard has been met.

kaitinghh commented on src/main/java/duke/Deadline.java:
```diff
@@ -0,0 +1,91 @@
+                    + "dd/mm/yyyy hhmm (e.g. 30/01/2024 1234");
+        }
+
+        switch (logic) {
```
Since you only have one case, would an if else be more concise than switch?

kaitinghh commented on src/main/java/duke/DukeException.java:
```diff
@@ -0,0 +1,14 @@
+/**
+ * An extension of the Exception class used to identify exceptions unique to Duke.
+ */
+public class DukeException extends Exception {
```
Could there be a better naming for DukeException class?

LWS49 replied:
Could it be better to have different types of DukeException instead? With this way of implementation all error messages would be in this file, and can be edited without going into other files unnecessarily. (encapsulation)

kaitinghh commented on src/main/java/duke/Storage.java:
```diff
@@ -0,0 +1,118 @@
+ */
+public class Storage {
+
+    private final String dataPath = System.getProperty("user.dir") + "/data";
```
Final fields should be all capsed?

1rbg replied:
final and constant are not the same thing though, typically constants refer to class members, final just means it cannot be modified after instantiation. I think it's fine that they are not all capitalised since they are instance fields, however perhaps if the variables are class members, and then they should be in all caps.

kaitinghh commented on src/main/java/duke/TaskList.java:
```diff
@@ -0,0 +1,93 @@
+public class TaskList {
+    private String horzLine = "____________________________________________________________";
+
+    private ArrayList<Task> taskList = new ArrayList<>(100);
```
Why do you have to initialise the size of the taskList if its an arrayList?

kaitinghh reviewed (commented):
LGTM! Some minor comments for your review.

saiutkarsh33 commented:
LGTM!

1rbg commented on src/main/java/duke/Duke.java:
The Ui class should handle all UI-related functions, which include the printing of messages and the formatting to the terminal. Should all these functionality be moved to the Ui class?

1rbg commented on src/main/java/duke/ToDo.java:
```diff
@@ -0,0 +1,46 @@
+     * @param description Title of the task.
+     * @throws DukeException If description is empty.
+     */
+    public ToDo(String logic, String description) throws DukeException {
```
Using `logic` as a variable name might not be very clear as it is not apparent what it is, but naming it with an 'is_' prefix is also not right as it isn't exactly a boolean, maybe you could rename this to `marked` or something similar, just a suggestion.

1rbg commented on src/main/java/duke/ToDo.java:
```diff
@@ -0,0 +1,46 @@
+        if (description.trim().length() == 0) {
+            throw new DukeException("Please enter a description for this todo task");
+        }
+        switch (logic) {
+        case "1":
+            this.isDone = true;
+            break;
+        default:
+            this.isDone = false;
+            break;
+        }
```
If there are only 2 cases to be handled here, "1", or everything else, could using an if statement be better?

1rbg reviewed (commented):
I like that your code is pretty well organised, except for maybe the Ui functionality still residing in Duke. I have some minor comments about the code quality, and provided some suggestions. Overall well done so far!

LWS49 commented on src/main/java/duke/Deadline.java:
```diff
@@ -0,0 +1,91 @@
+    public Deadline(String description, String time) throws DukeException {
+        super(description);
+        time = time.trim();
+        if (description.trim().length() == 0 || time.length() == 0) {
+            String errorMessage = "Please enter the following:";
+            if (description.trim().length() == 0) {
+                errorMessage += "\na description for this duke.Deadline task";
+            }
+            if (time.length() == 0) {
+                errorMessage += "\na deadline after the '/by' command";
+            }
```
That's a good way to customize your error message!

LWS49 commented on src/main/java/duke/TaskList.java:
```diff
@@ -0,0 +1,93 @@
+        }
+        Task tempTask = taskList.get(index - 1);
+        taskList.remove(index - 1);
+        printWithLines("Noted. I've removed this task:\n   " + tempTask.toString()
+                + "\nNow you have " + taskList.size() + (taskList.size() > 1 ? " tasks " : " task ")
+                + "in the list.");
+    }
```
this should be under UI, right? Since UI should be handling all outputs.

jayllo-c committed:
Switch GUI to JavaFX

Now uses JavaFX to interact with the user.

Added Features:
* The user will now see a pop up window with message boxes to interact
with destiny.
* The user will have a dedicated text box to type in.
* The user interface can be edited through Scene Builder

Other changes:
* Normal UI no longer available

This is done to make the chatbot more useable and visually appealing.

jayllo-c committed:
Merge branch 'branch-Level-10'

jayllo-c committed:
Improve coding standard

Used check style to ensure that there are no detectable coding
standard violations.

This was done to ensure the minimum standard is adhered to.

jayllo-c committed:
Bug Fix: Save after edit to tasks

Restructured the Destiny class slightly to ensure that the save command
executes upon any change to the task list.

jayllo-c committed:
Bug fix: fix refactor issues

Refactoring the code to fit JavaFX has caused minor flaws in the text of
error messages generated. For example, the error message may indicate
the wrong function, which would confuse users.

Fixing this will make it easier for the users to understand why
the errors occur.

jayllo-c committed:
no message

jayllo-c committed:
Add assertions

Added assertions to the Task.java file.

These assertions were added to ensure that the tasks reflected
if the were marked or unmarked correctly.

jayllo-c committed:
Improve code quality

Shifted the functionality of the commands out of the Destiny class
and into individual command classes.

New classes created:
* Command (abstract)
* DeadlineCmd
* DeleteCmd
* EventCmd
* FindCmd
* ListCmd
* MarkDoneCmd
* MarkNotDoneCmd
* ToDoCmd

Changes have also been made to the parser class to accommodate the
shift in functionality.

All this was done to shorten the length of the getResponse method in
Destiny because it was too long.

jayllo-c committed:
Merge pull request #2 from jayllo-c/branch-A-CodeQuality

Improve code quality

jayllo-c committed:
Merge branch 'master' into branch-A-Assertions

jayllo-c committed:
Merge pull request #3 from jayllo-c/branch-A-Assertions

Add assertions

jayllo-c committed:
Detect Duplicates

Add the ability to recognize and deal with duplicate items.
e.g., the same task added multiple times.

This was done by adding the equals() method to the task class,
and checking in the addTask() method of the TaskList to see if the
new task is a duplicate of any of the tasks already in the list.
If duplicate is detected, it is not added to the list.

This is to ensure that multiple tasks of the same entry are not added
to the TaskList since they would be redundant.

jayllo-c committed:
Merge branch 'branch-BCD-Extension'

jayllo-c committed:
Improve GUI

Improved the GUI by changing the background, the user profile pictures
and the shape of each dialog box.

This was done to improve the user experience.

jayllo-c committed:
Translucent DialogBox

Made dialog boxes translucent to enhance user experience.

jayllo-c committed:
Implement GUI exit

GUI now exits upon receiving the command bye after a 5 second delay.

jayllo-c committed:
Bug Fix: bye cmd

Previously reached error when bye was typed without a space at the back.

Now works as intended.

jayllo-c committed:
Finish User Guide

User guide has been create, which will help users understand how Destiny
works.

jayllo-c committed:
add _config.yml

This is a config file in the docs directory to create a custom theme
for the project website.

jayllo-c committed:
Update _config.yml

jayllo-c committed:
Improve code quality

General improvements to code quality were made to ensure that the
code adheres to a high standard.

jayllo-c committed:
Merge branch 'master' of https://github.com/jayllo-c/ip

jayllo-c committed:
Parser.java: Improve code quality

Renamed Boolean variable to ensure that the
code adheres to a high coding standard.

jayllo-c committed:
DialogBox.java: Improve code quality

Reduced method length to ensure that the
code adheres to a high coding standard.

jayllo-c committed:
More automated tests

Added testing for list command and todo tasks, and tested task classes
more comprehensively.

This is to better ensure the working integrity of the code.

jayllo-c committed:
Fix Sample Data

Rectified the sample data by removing unwanted entries so as to
prevent confusion for users of the chatbot.

jayllo-c committed:
Edit docs

Slight changes were made to alter the user guide page.

jayllo-c committed:
Edit docs

Slight changes were made to alter the user guide page.

jayllo-c committed:
Edit docs

Slight changes were made to alter the user guide page.

jayllo-c committed:
Edit docs

Slight changes were made to alter the user guide page.

jayllo-c committed:
Edit docs

Slight changes were made to alter the user guide page.

jayllo-c committed:
Edit docs

Slight changes were made to alter the user guide page.
